### PR TITLE
Ensure that javax.websocket Encoder/Decoder destroy method is called.

### PR DESCRIPTION
**Pull request by lachlan-roberts**
Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

### CODE REVIEW

File: jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/RegisteredDecoder.java [hunk 1]
```diff
@@ -74,6 +78,23 @@ public boolean isType(Class<?> type)
         return (T)instance;
     }
 
+    public void destroyInstance()
```
Comment (sbordet) [inline jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/RegisteredDecoder.java:81]: We have an interface `Destroyable`, any reason to not use it?

File: jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/AvailableDecoders.java [hunk 2]
```diff
@@ -207,4 +207,9 @@ public RegisteredDecoder getFirstRegisteredDecoder(Class<?> type)
     {
         return registeredDecoders.stream();
     }
+
+    public void close()
```
Comment (sbordet) [inline jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/AvailableDecoders.java:213]: Implement `Closeable` then?

File: jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/encoders/AvailableEncoders.java [hunk 3]
```diff
@@ -62,6 +67,23 @@ public boolean isType(Class<?> type)
             return objectType.isAssignableFrom(type);
         }
 
+        public void destroyInstance()
```
Comment (sbordet) [inline jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/encoders/AvailableEncoders.java:71]: Ditto

File: jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/encoders/AvailableEncoders.java [hunk 4]
```diff
@@ -286,4 +308,9 @@ public boolean test(Class<?> type)
     {
         return registeredEncoders.stream().anyMatch(registered -> registered.isType(type));
     }
+
+    public void close()
```
Comment (sbordet) [inline jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/encoders/AvailableEncoders.java:314]: Ditto

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/RegisteredDecoder.java (hunk 1 above)
Comment (joakime) [reply jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/RegisteredDecoder.java:81]: Is the `Destroyable` interface exposed to the WebApp?

Comment (lachlan-roberts) [reply jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/RegisteredDecoder.java:81]: > Is the Destroyable interface exposed to the WebApp?

I don't think it needs to be. `RegisteredDecoder` is an internal implementation anyway not something used by the WebApp.

Comment (lachlan-roberts) [reply jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/RegisteredDecoder.java:81]: I don't think there is any benefit from implementing `Destroyable` though, we don't really need the JMX operations for destroy on this.

This usage doesn't really line up with the description of `Destroyable` from the javadoc:
> Typically a Destroyable is a {@link LifeCycle} component that can hold onto
>  resources over multiple start/stop cycles. A call to destroy will release all
>  resources and will prevent any further start/stop cycles from being successful.

Comment (sbordet) [reply jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/RegisteredDecoder.java:81]: Fair enough.

**Review by sbordet — APPROVED** [review]

**Review by joakime — APPROVED** [review]

**Pull request merged**